File: perf-test/src/gatling/simulations/heap/HeapSimulation.scala
package heap

import java.time.Duration

import io.gatling.core.Predef._
import io.gatling.http.HeaderValues._
import io.gatling.http.Predef._
import io.gatling.http.protocol.HttpProtocolBuilder

/**
 * @author Jonatan Ivanov
 */
class HeapSimulation extends Simulation  {
    val size = "512KB"
    val ttl: Duration = Duration.ofMinutes(1)
    val rate = 10 // 5MB/sec (300MB/min), 10 objects/sec (600 objects/min)
    val duration = 300

    println(s"size: $size")
    println(s"ttl: $ttl")
    println(s"rate: $rate")
    println(s"duration: $duration")

    val httpProtocol: HttpProtocolBuilder = http
        .baseUrl("http://localhost:8080")
        .contentTypeHeader(ApplicationJson)
        .acceptHeader(ApplicationJson)

    setUp(
        scenario("heapStress")
            .exec(
                http("createObject")
                    .post("/resources/objects")
                    .body(StringBody(s"""{ "size": "$size", "ttl": "$ttl" }""")).asJson
            )
            .inject(constantUsersPerSec(rate) during duration)
            .protocols(httpProtocol)
    )
}

File: perf-test/src/gatling/java/resourceater/simulations/heap/HeapSimulation.java
package resourceater.simulations.heap;

import java.time.Duration;

import io.gatling.javaapi.core.ChainBuilder;
import io.gatling.javaapi.core.Simulation;
import io.gatling.javaapi.http.HttpProtocolBuilder;

import static io.gatling.http.HeaderNames.Accept;
import static io.gatling.http.HeaderNames.ContentType;
import static io.gatling.http.HeaderValues.ApplicationJson;
import static io.gatling.javaapi.core.CoreDsl.StringBody;
import static io.gatling.javaapi.core.CoreDsl.constantUsersPerSec;
import static io.gatling.javaapi.core.CoreDsl.exec;
import static io.gatling.javaapi.core.CoreDsl.scenario;
import static io.gatling.javaapi.http.HttpDsl.http;

@SuppressWarnings("unused")
public class HeapSimulation extends Simulation {
    final String size = "512KB";
    final Duration ttl = Duration.ofMinutes(1);
    final int usersPerSec = 10; // 5MiB/sec (300MiB/min), 10 objects/sec (600 objects/min)
    final Duration duration = Duration.ofSeconds(300);

    final HttpProtocolBuilder httpProtocol = http.baseUrl("http://localhost:8080")
        .contentTypeHeader(ApplicationJson())
        .acceptHeader(ApplicationJson());

    final ChainBuilder createObject = exec(http("createObject")
        .post("/resources/objects")
        .header(ContentType(), ApplicationJson())
        .header(Accept(), ApplicationJson())
        .body(StringBody("{ \"size\": \"%s\", \"ttl\": \"%s\" }".formatted(size, ttl))).asJson()
    );

    {
        System.out.printf("size: %s%n", size);
        System.out.printf("ttl: %s%n", ttl);
        System.out.printf("usersPerSec: %s%n", usersPerSec);
        System.out.printf("duration: %s%n", duration);

        setUp(scenario("heapStress")
            .exec(createObject)
            .injectOpen(constantUsersPerSec(usersPerSec).during(duration))
        ).protocols(httpProtocol);
    }
}
